tss/rsa/rsa_threshold.go

```diff
@@ -0,0 +1,251 @@
+	pprime.Sub(p, ONE)
+
+	// q - 1
+	var m big.Int
+	m.Sub(q, ONE)
+	// (p - 1)(q - 1)
+	m.Mul(&m, &pprime)
+	// >> 2 == / 4
+	m.Rsh(&m, 2)
+
+	// de ≡ 1
+	var d big.Int
+	_d := d.ModInverse(big.NewInt(e), &m)
+
+	if _d == nil {
+		return nil, errors.New("rsa_threshold: no ModInverse for e in Z/Zm")
+	}
+
+	// a_0...a_{k-1}
+	a := make([]*big.Int, threshold)
+	// a_0 = d
+	a[0] = &d
+
+	// a_0...a_{k-1} = rand from {0, ..., m - 1}
+	for i := int64(1); i <= threshold-1; i++ {
+		a[i], err = rand.Int(randSource, &m)
+		if err != nil {
+			return nil, errors.New("rsa_threshold: unable to generate an int within [0, m)")
+		}
+	}
+
+	shares := make([]KeyShare, players)
+
+	// 1 <= i <= l
+	for i := int64(1); i <= players; i++ {
+		// Σ^{k-1}_{i=0} | a_i * X^i (mod m)
+		poly := computePolynomial(threshold, a, i, &m)
+		shares[i-1].si = poly
+		shares[i-1].Index = uint8(i)
+		if cache {
+			shares[i-1].get2DeltaSi(players)
+		}
+	}
+
+	return shares, nil
+}
+
+func calcN(p, q *big.Int) big.Int {
+	// n = pq
+	var n big.Int
+	n.Mul(p, q)
+	return n
+}
+
+// f(X) = Σ^{k-1}_{i=0} | a_i * X^i (mod m)
+func computePolynomial(k int64, a []*big.Int, x int64, m *big.Int) *big.Int {
+	sum := big.NewInt(0)
+
+	//  Σ^{k-1}_{i=0}
+	for i := int64(0); i <= k-1; i++ {
+		// X^i
+		// TODO optimize: we can compute x^{n+1} from the previous x^n
+		xi := int64(math.Pow(float64(x), float64(i)))
+		// a_i * X^i
+		prod := big.Int{}
+		prod.Mul(a[i], big.NewInt(xi))
+		// (mod m)
+		prod.Mod(&prod, m) // while not in the spec, we are eventually modding m, so we can mod here for efficiency
+		// Σ
+		sum.Add(sum, &prod)
+	}
+
+	sum.Mod(sum, m)
+
+	return sum
+}
+
+// PadHash MUST be called before signing a message
+func PadHash(padder Padder, hash crypto.Hash, pub *rsa.PublicKey, msg []byte) ([]byte, error) {
+	// Sign(Pad(Hash(M)))
+
+	hasher := hash.New()
+	hasher.Write(msg)
+	digest := hasher.Sum(nil)
+
+	return padder.Pad(pub, hash, digest)
+}
+
+type Signature []byte
+
+// CombineSignShares combines t SignShare's to produce a valid signature
+func CombineSignShares(players int64, pub *rsa.PublicKey, shares []SignShare, msg []byte) (Signature, error) {
+	w := big.NewInt(1)
+	delta := calculateDelta(players)
+	// i_1 ... i_k
+	for _, share := range shares {
+		// λ(S, 0, i)
+		lambda, err := computeLambda(delta, shares, 0, int64(share.Index))
+		if err != nil {
+			return nil, err
+		}
+		// 2λ
+		var exp big.Int
+		exp.Add(lambda, lambda) // faster than TWO * lambda
+
+		// we need to handle negative λ's (aka inverse), so abs it, compare, and if necessary modinverse
+		abslam := big.Int{}
+		abslam.Abs(&exp)
+		var tmp big.Int
+		// x_i^{|2λ|}
+		tmp.Exp(share.xi, &abslam, pub.N)
+		if abslam.Cmp(&exp) == 1 {
+			tmp.ModInverse(&tmp, pub.N)
+		}
+		// TODO  first compute all the powers for the negative exponents (but don't invert yet); multiply these together and then invert all at once. This is ok since (ab)^-1 = a^-1 b^-1
+
+		w.Mul(w, &tmp)
+	}
+	w.Mod(w, pub.N)
```

Comment:
need to add modular reduction, otherwise numbers grow fast.
```suggestion
		w.Mul(w, &tmp).Mod(w, pub.N)
	}

```